README.md

```diff
@@ -25,16 +25,44 @@ Once you have implemented the `sa` program (and tested it to the best of your ab
 
+For the search part we have implemented a binary search function that finds upper bounds and lower bounds. Our variant of binary search runs O(m * log(n)). This is the running time for both finding lower and upper bound. We start with array size n and divide search space by two each iteration, thus a total of O(log n) iterations. In each iteration we compare the pattern m, taking up to O(m).  
+This of course only the time it takes to return two indexes. The time it would take to get the actual results in SAM or similar would obviously be bounded by z.
 ## Insights you may have had while implementing the algorithm
 
+One of the most interesting insights was the great difference in running time for construction vs searching.
+This also made it a bit complicated to validate the searching part, since long search times were hard to construct.
+
 ## Problems encountered if any
 
+We had some troubles identifiying the different edge cases for the binary search algorithm more specifically the upper and lower bound algorithms. 
+
 ## Validation
 
-*How did you validate that the construct and the search algorithm works?*
+We created some suffix arrays and did binary search on them. 
+
+First we tested the results on some specific and complex strings in order to see if some edgecases did not work. 
+The resulting interval of matches were then checked checked. For suffixes outside the interval we checked that they were not a match, and for values in the interval we obviously checked that they indeed were a match.
+
+Second we compared the results with results from the naive algorithm in project 1. Here we compared two SAM files with the respective outputs.
+
+Last we made a test method to generate many random strings over different alphabets (english, all a's, ACTG) and checked again that all values in interval and outside interval was placed correctly.
 
 ## Running time
 
-*List experiments and results that show that both the construction algorithm and the search algorithm works in the expected running time. Add figures by embedding them here, as you learned how to do in project 1.*
+Construction of the algorithm only depends on the size of our genome or string. If we plot time/n on the y-axis and n on the x-axis we get a linear correlation. From the image the correlation seems pretty clear.
+![](figs/construction_time_pr3.png)
+
+As mentioned we found it a bit complicated to test the search part of the algoritm.
+When plotting length of search pattern m on x-axis and time / log(n+z) on the y-axis we would expect a linear relationship.
+Since the running times are so fast compared to the time it takes to create the suffix array the measured times are very low which means we have a lot of fluctuations:
+
+![](figs/search_time_pr3_n=50k.png)
```

Comment:
It shouldn't be $\log(n+z)$. The output contribution isn't logarithmic. The running time is $O(m\log n + z)$.

I would ignore $z$ here (I can always test it later) and just measure the time it takes to find the intervals, not output them. Then the time should be $O(m\log n)$, and while it isn't easy to get a single plot that shows this running time, it isn't hard to do it with multiple plots or colours. It isn't substantially more difficult to do than when you had to plot the $O(nm)$ running time in the first project.

But it looks fine.